haas/api.py

```diff
@@ -133,7 +133,65 @@ def project_detach_node(project, node):
     project.nodes.remove(node)
     db.session.commit()
 
-
+@rest_call('PUT', '/network/<network>/access/<project>')
+def network_grant_project_access(project, network):
+   """Add access to <network> to <project>.
+
+   If the project or network does not exist, a NotFoundError will be raised.
+   If the project already has access to the network a DuplicateError will be raised.
+   """
+   network = _must_find(model.Network, network)
+   project = _must_find(model.Project, project)
+   #must be admin or the owner of the network to add projects
+   get_auth_backend().require_project_access(network.owner)
+
+   if project in network.access:
+       raise DuplicateError('Network %s is already in project %s'%
+                            (network.label, project.label))
+
+   network.access.append(project)
+   db.session.commit()
+
+@rest_call('DELETE', '/network/<network>/access/<project>')
+def network_revoke_project_access(project, network):
+    """Remove access to <network> from <project>.
+
+    If the project or network does not exist, a NotFoundError will be raised.
+    If the project is the owner of the network a BlockedError will be raised.
+    """
+    network = _must_find(model.Network, network)
+    project = _must_find(model.Project, project)
+    #must be admin, the owner of the network, or <project> to remove <project>.
+    authorized = get_auth_backend().have_project_access(network.owner)
+    
+    if network.access:
+        for proj in network.access:
```

Comment:
What about this (with fixed up indentation and whatnot)?

```python
owner = network.owner

# must be admin, the owner of the network, or <project> to remove <project>. 
if not ( auth_backend().have_admin() or \
           (owner is not Null and auth_backend.have_project_access(owner)) or \
           (project in network.access and auth_backend.have_project_access(project)):
    raise ...
```

Note that `network.owner` can be `None`, so we need to check for that (ie - the `authorized` check above is not valid). As a side note, I just opened #623 to allow `None` to be passed, since that seems like a reasonable way to simplify the code.